Discards, tuples and out variables

MadsTorgersen opened the pull request:
This is a very rough cut of discards, tuples and out variables, and is here for discussion/sharing purposes. It lacks examples and cross references throughout and doesn't touch on patterns at all. It's also missing treatment of additional declaration spaces at statement/expression level, as well as the syntactic short form `var (a, b)` in deconstruction.

The intent is to test out an alternative approach to #44, #63 and #596 along the lines suggested in #561, and to fill out several gaps in those PRs.

The approach differs in several ways:

- Discards are proper variables, only without names. Otherwise, the treatment of out vars cannot explain where the variable comes from that is passed to the out parameter.
- The entities on the left hand side of a deconstructing assignment are proper tuple expressions. Otherwise we cannot adequately account for when their elements are full-blown expressions.
- A new expression classification for `tuples` helps describe when, whether and how tuples should be evaluated.
- The variable-declaring capability of tuples and out variables is factored out into a new kind of expression, a "declaration expression" instead of adding new special cases to each of their productions.

I believe this approach is more correct now, and more resilient to future language additions. But there is still a ways to go to clean it up.

MadsTorgersen committed:
Rough cut at discards, tuples and out variables

KalleOlaviNiemitalo commented on standard/conversions.md:
```diff
@@ -429,6 +439,10 @@ For an explicit reference conversion to succeed at run-time, the value of the so
 
+### §explicit-tuple-conversions Explicit tuple conversions
+
+An explicit conversion exists from a tuple expression `E` to a tuple type `T` if `E` has the same arity as `T` and an implicit or explicit conversion exists from each element in `E` to the corresponding element type in `T`. The conversion is performed by creating an instance of `T`'s corresonding `System.ValueTuple<...>` type, evaluating each tuple element expression of `E` in order, converting i ...
```
Typo "corresonding".

MadsTorgersen replied:
Thanks! Fixed.

KalleOlaviNiemitalo commented on standard/types.md:
```diff
@@ -378,6 +387,14 @@ No standard conversions exist between `bool` and other value types. In particula
+
+A tuple type represents an ordered fixed-length sequence of values with optional names and individual types. A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional. This syntax is shorthand for `System.ValueTuple<T1,..., Tn>` (or an equivalent type for long tuple types), which is a generic struct type available for each "arity" (leng ...
+
+The `ValueTuple<...>` types do not represent the optional element names, which are thus not part of the runtime representation of the tuple value, but impact how the tuple elements can be accessed. There is an identy conversion between all tuple types with the same arity and element types, and with the corresponding constructed `ValueTuple<...>` type.
```
Typo "identy".

MadsTorgersen replied:
Fixed.

KalleOlaviNiemitalo commented on standard/variables.md:
```diff
@@ -150,6 +150,10 @@ A local variable introduced by a *local_variable_declaration* is not automatical
 
+#### 9.2.8.1 Discards
+
+A ***discard*** is a local variable that has no name, and can thus only be used once, at the point where it is introduced. A discard is not initially assigned, so it is always an error to access its the value. A discard is introduced by a *declaration_expression* or *tuple_deconstruction* with the identifier `_`, and by use of the *simple_name* `_` when lookup of that name does not yield a result ...
```
OK, the *simple_name* case appears to cover the following, so the *declaration_expression* syntax does not need to allow it.

```C#
public abstract class C {
    protected abstract void O(out int i);
    public void M() {
        O(out _);
    }
}
```

MadsTorgersen replied:
Indeed. That is why a *declaration_expression* requires a *local_variable_type*. So there's no overlap, right?

ericlippert replied:
"its the value" looks like a misedit.

MadsTorgersen replied:
Indeed. Since fixed. Thanks!

MadsTorgersen committed:
Fix typos

MadsTorgersen committed:
Fix indentation and whitespace

jskeet commented on standard/conversions.md:
```diff
@@ -77,7 +78,11 @@ However, dynamic conversions ([§10.2.10](conversions.md#10210-implicit-dynamic-
+
+- between `object` and `dynamic`.
+- between tuple types with the same arity, when an identity conversion exists between each pair of corresponding element types.
+- between constructed types that are the same when replacing constituent types with identity convertible ones.
```
This last bullet is saying there's an identity conversion between `List<dynamic>` and `List<object>` etc, right?

MadsTorgersen replied:
Correct. Examples would be helpful of course! I'll get there.

jskeet commented on standard/expressions.md:
```diff
@@ -14,6 +14,7 @@ The result of an expression is classified as one of the following:
 - A variable. Every variable has an associated type, namely the declared type of the variable.
 - A null literal. An expression with this classification can be implicitly converted to a reference type or nullable value type.
 - An anonymous function. An expression with this classification can be implicitly converted to a compatible delegate type or expression tree type.
+- A tuple. Every tuple has a fixed number of elements, each with an expression and an optional tuple element name.
```
Each with an expression, or each with a value?

jskeet replied:
Nope, I'm wrong :)

jskeet commented on standard/expressions.md:
```diff
@@ -37,6 +38,7 @@ Most of the constructs that involve an expression ultimately require the express
 - The value of a variable is simply the value currently stored in the storage location identified by the variable. A variable shall be considered definitely assigned ([§9.4](variables.md#94-definite-assignment)) before its value can be obtained, or otherwise a compile-time error occurs.
 - The value of a property access expression is obtained by invoking the *get_accessor* of the property. If the property has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed, and the result of the invocation becomes the value of the property access expression.
 - The value of an indexer access expression is obtained by invoking the *get_accessor* of the indexer. If the indexer has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed with the argument list associated with the indexer access expression, and the result of the invocation becomes the v ...
+- The value of a tuple is obtained by applying an implicit tuple conversion to a receiving tuple type, if one exists, or to the type of the tuple expression itself if not.
```
Can we go over this in the meeting? I'm not following at the moment.

MadsTorgersen replied:
Sure!

MadsTorgersen committed:
Fix up tuples

jskeet commented on standard/expressions.md:
```diff
@@ -4300,6 +4322,27 @@ A *throw expression* shall only occur in the following syntactic contexts:
+- As an `out` *argument_value* in an *argument_list*.
+- As a *tuple_element* in a *tuple_expression* that occurs on the left-hand side of a deconstructing assignment.
+
+The *local_variable_type* of a *local_variable_declaration* either directly specifies the type of the variable introduced by the declaration, or indicates with the identifier `var` that the type is implicit and should be inferred based on the syntactic context as follows:
```
Should *local_variable_declaration* here be *declaration_expression*?

MadsTorgersen replied:
Yes. Thanks!

MadsTorgersen replied:
Fixed!

jskeet reviewed (commented):
Submitting the comments as far as I've got. (More reading and thinking to do.)

MadsTorgersen committed:
Fix section and word converter errors

MadsTorgersen committed:
Fix typos and lint errors

MadsTorgersen committed:
Merge branch 'draft-v7' into Mads-v7-local-declarations-in-expressions

MadsTorgersen committed:
Crisp up deconstruction

Make sure evaluation order is described precisely in deconstructing assignments.

BillWagner commented on standard/variables.md:
```diff
@@ -150,6 +150,10 @@ A local variable introduced by a *local_variable_declaration* is not automatical
 
+#### §discards-new-clause Discards
+
+A ***discard*** is a local variable that has no name, and can thus only be used once, at the point where it is introduced. A discard is not initially assigned, so it is always an error to access its value. A discard is introduced by a *declaration_expression* or *tuple_deconstruction* with the identifier `_`, and by use of the *simple_name* `_` when lookup of that name does not yield a result ().
```
I'm curious about the parentheses at the end of this paragraph.

MadsTorgersen replied:
It was a reminder to add a section reference. Fixed in upcoming commit.

BillWagner commented on standard/types.md:
```diff
@@ -378,6 +387,18 @@ No standard conversions exist between `bool` and other value types. In particula
 
+### §tuple-types-new-clause Tuple types
+
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...T ...
```
nit: space missing.

```suggestion
A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***. A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...Tn` from `System.ValueTuple<...>`, which shall be a set of generic struct types capable of expressing every arity of tuple types.
```

BillWagner commented on standard/types.md:
```diff
@@ -378,6 +387,18 @@ No standard conversions exist between `bool` and other value types. In particula
+
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...T ...
+
+Element names within a tuple type shall be distinct. It is an error for an explicit tuple element name to be of the form `ItemX` where `X` is any sequence of non-`0`-initiated decimal digits that could represent the position of a tuple element, except as the name for that actual element.
```
As we finish this approach, I think an example for this would add clarity.

MadsTorgersen replied:
fixed in upcoming commit

BillWagner commented on standard/types.md:
```diff
@@ -378,6 +387,18 @@ No standard conversions exist between `bool` and other value types. In particula
+
+Element names within a tuple type shall be distinct. It is an error for an explicit tuple element name to be of the form `ItemX` where `X` is any sequence of non-`0`-initiated decimal digits that could represent the position of a tuple element, except as the name for that actual element.
+
+The optional element names are not represented in the `ValueTuple<...>` types, and are not stored in the runtime representation of a tuple value. There is an identity conversion between all tuple types with the same arity and element types, as well as to and from the corresponding constructed `ValueTuple<...>` type.
```
Should we add that the element types must be in the same order for the identity conversion?

MadsTorgersen replied:
Fixed in upcoming commit

BillWagner commented on standard/conversions.md:
```diff
@@ -77,7 +78,11 @@ However, dynamic conversions ([§10.2.10](conversions.md#10210-implicit-dynamic-
+
+- between `object` and `dynamic`.
+- between tuple types with the same arity, when an identity conversion exists between each pair of corresponding element types.
+- between constructed types that are the same when replacing constituent types with identity convertible ones.
```
Possible alternative phrasing:

```suggestion
- between constructed types built from the same generic type where there exists an identity conversion between each corresponding type argument.
```

MadsTorgersen replied:
Adopted a variant of your phrasing in upcoming commit

BillWagner commented on standard/expressions.md:
```diff
@@ -37,6 +38,7 @@ Most of the constructs that involve an expression ultimately require the express
 - The value of a variable is simply the value currently stored in the storage location identified by the variable. A variable shall be considered definitely assigned ([§9.4](variables.md#94-definite-assignment)) before its value can be obtained, or otherwise a compile-time error occurs.
 - The value of a property access expression is obtained by invoking the *get_accessor* of the property. If the property has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed, and the result of the invocation becomes the value of the property access expression.
 - The value of an indexer access expression is obtained by invoking the *get_accessor* of the indexer. If the indexer has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed with the argument list associated with the indexer access expression, and the result of the invocation becomes the v ...
+- The value of a tuple expression is obtained by applying an implicit tuple conversion (§implicit-tuple-conversions-new-clause) to the type of the tuple expression itself. It is an error to obtain the value of a tuple expression that does not have a type.
```
Is the last sentence necessary? Can a tuple expression ever not have a type?

MadsTorgersen replied:
Yes, it is necessary. A tuple expression may not have a type if some of its elements do not have a type, e.g. `(null, s => s)`.

BillWagner commented on standard/variables.md:
```diff
@@ -150,6 +150,10 @@ A local variable introduced by a *local_variable_declaration* is not automatical
+#### §discards-new-clause Discards
+
+A ***discard*** is a local variable that has no name, and can thus only be used once, at the point where it is introduced. A discard is not initially assigned, so it is always an error to access its value. A discard is introduced by a *declaration_expression* or *tuple_deconstruction* with the identifier `_`, and by use of the *simple_name* `_` when lookup of that name does not yield a result ().
+
```
Somewhere here, we need to make sure that multiple discards are allowed in the same scope:

```csharp

_ = M1();  // Discard's type is int
_ = M2();  // Discard's type is string

int M1() => 1;
string M2() => "hello";
```

MadsTorgersen replied:
I believe the treatment of discards in expressions.md (Simple Names and Declaration Expressions) takes care of that by ensuring they do not cause anything to be declared or added to the declaration space.

BillWagner commented on standard/expressions.md:
```diff
@@ -1456,6 +1458,11 @@ A *simple_name* is either of the form `I` or of the form `I<A₁, ..., Aₑ>`,
     - Otherwise, if the namespaces imported by the *using_namespace_directive*s of the namespace declaration contain exactly one type having name `I` and `e` type parameters, then the *simple_name* refers to that type constructed with the given type arguments.
     - Otherwise, if the namespaces imported by the *using_namespace_directive*s of the namespace declaration contain more than one type having name `I` and `e` type parameters, then the *simple_name* is ambiguous and a compile-time error occurs.  
   > *Note*: This entire step is exactly parallel to the corresponding step in the processing of a *namespace_or_type_name* ([§7.8](basic-concepts.md#78-namespace-and-type-names)). *end note*
+- Otherwise, if `e` is zero and `I` is the identifier `_`, the *simple_name* is a discard () and its type shall be inferred from the syntactic context:
```
See my comment at the end of variables. I Think this wording disallows multiple discards in the same scope.

MadsTorgersen replied:
Can you elaborate what gives you that impression? I feel like this wording very explicitly gives each discard its own type.

For now I'm adding the word "fresh" ("the *simple_name* is a fresh discard") and a section reference to Discards in variables.md, and mark this resolved, but please reopen if that doesn't address your concern.

BillWagner commented on standard/expressions.md:
````diff
@@ -1470,6 +1477,23 @@ parenthesized_expression
+    : (identifier ':')? expression
+```
+
+A tuple expression is classified as a tuple. It has a type if all the element expressions have a type. In that case, the type of a tuple expression is the tuple type of the same arity, where each type element has the type of the corresponding element expression and the same identifier as the corresponding tuple element if it has one, or none if it does not.
````
grammar concern:  I find this sentence hard to parse correctly. Does "or none if it does not" apply to the names of tuple elements, the type of each element, or the tuple itself?

MadsTorgersen replied:
Reworded in upcoming commit to "... where each type element has the type - and name, where given - of the corresponding element expression."

BillWagner commented on standard/expressions.md:
```diff
@@ -5901,6 +5948,26 @@ When a property or indexer declared in a *struct_type* is the target of an assig
+  - Otherwise, if `E` has a tuple type (§tuple-types-new-clause) with the same arity as the left-hand side, then `E1, E2, ...` shall be the member access expressions `E.Item1, E.Item2, ...`, except that `E` shall be evaluated only once.
+  - Otherwise, if `E.Deconstruct(out var v1, out var v2, ...)` with the number of arguments corresponding to the arity of the left hand side is a valid instance method invocation, then that invocation shall be performed and `E1, E2, ...` shall be the simple name expressions `v1, v2, ...`.
+  - Otherwise `E` cannot be deconstructed, and a compile time error occurs.
+- Finally, the operation is evaluated as `(R1 = E1, R2 = E2, ...)`, except that `R1, R2, ...` have already been evaluated and are not evaluated again. This means that individual element-wise assignments are recursively performed, and a resulting tuple value is created.
```
Should this be `E1, E2, ...`? If I understand correctly that point is to make sure that `(x, y) = (y, x)` works correctly.

```suggestion
- Finally, the operation is evaluated as `(R1 = E1, R2 = E2, ...)`, except that `E1, E2, ...` have already been evaluated and are not evaluated again. This means that individual element-wise assignments are recursively performed, and a resulting tuple value is created.
```

MadsTorgersen replied:
You are right that evaluation of `E1, E2, ...` happens to late here. In upcoming commit I've inserted a step that evaluates and converts each `E1, E2, ...` , and adjusted the final step to say that those also are not evaluated again.

BillWagner commented on standard/expressions.md:
```diff
@@ -5901,6 +5948,26 @@ When a property or indexer declared in a *struct_type* is the target of an assig
+  - Otherwise `E` cannot be deconstructed, and a compile time error occurs.
+- Finally, the operation is evaluated as `(R1 = E1, R2 = E2, ...)`, except that `R1, R2, ...` have already been evaluated and are not evaluated again. This means that individual element-wise assignments are recursively performed, and a resulting tuple value is created.
+
+Note that even though the assignment operation "evaluated as" a tuple expression containing element-wise assignment expressions, this does prevent the deconstructing assignment from occurring as a statement expression, nor the elements on the left-hand side from being declaration expressions.
```
I'm not sure I understand this. Should it be "this doesn't prevent..." or should the "nor" be an "or"?

MadsTorgersen replied:
The former. Fixed in upcoming commit.

BillWagner reviewed (commented):
I like this approach. I likely went into details that were beyond where you've already speced things out, but I wanted to make sure I noted them, even if just to extend the TODO list.

This does make a number of concepts more cohesive.

MadsTorgersen committed:
Merge branch 'draft-v7' into Mads-v7-local-declarations-in-expressions

MadsTorgersen committed:
Apply suggestions from code review

Co-authored-by: Bill Wagner <wiwagn@microsoft.com>

MadsTorgersen committed:
Add examples and section references, fix issues from review comments

MadsTorgersen committed:
Merge branch 'Mads-v7-local-declarations-in-expressions' of https://github.com/dotnet/csharpstandard into Mads-v7-local-declarations-in-expressions

MadsTorgersen committed:
Fix lint issues

MadsTorgersen commented:
Current status - the PR is taking shape but the following work is still TODO:

- Examples of Declaration expressions, Discards, Tuple expressions and Deconstructing assignments
- Add declaration_expression to Declaration Spaces where necessary
- Generalize scope to not just be blocks where necessary (to allow expression bodies with declaration expressions)

The scope of this PR will not include patterns.

MadsTorgersen committed:
Add examples

MadsTorgersen committed:
Update variables.md

KalleOlaviNiemitalo commented on standard/expressions.md:
```diff
@@ -5984,6 +6058,27 @@ When a property or indexer declared in a *struct_type* is the target of an assig
+- Then the right hand side `E` is deconstructed into a list of expressions `E1, E2, ...` as follows:
+  - If `E` is a tuple expression with the same arity as the left-hand side, then `E1, E2, ...` shall be the element expressions of `E`.
+  - Otherwise, if `E` has a tuple type (§tuple-types-new-clause) with the same arity as the left-hand side, then `E1, E2, ...` shall be the member access expressions `E.Item1, E.Item2, ...`, except that `E` shall be evaluated only once.
+  - Otherwise, if `E.Deconstruct(out var v1, out var v2, ...)` with the number of arguments corresponding to the arity of the left hand side is a valid instance method invocation, then that invocation shall be performed and `E1, E2, ...` shall be the simple name expressions `v1, v2, ...`.
```
Should allow extension method invocations, too.

```csharp

static class C {
    static void M() {
        int a, b;
        (a, b) = default(D);
    }
    
    static void Deconstruct(this D d, out int a, out int b) {
        a = b = 0;
    }
}

struct D {}
```

[SharpLab](https://sharplab.io/#v2:CYLg1APgAgjAbAAigJgQYQQbwLACgEFLxIAsCAsgBQCUWehDCAlgHYAuCAhgDQIBGAbnqMClHv1oBeBMACmAM04BXADZtKAEWpD8hAL7CChooihkNsgMYB7FgGc2AJyWX1bABZM7CDTN7WlDlYOcQCg9gk6XRFOBGk+OIQABh0GA1x0vAdnVx8sPSA==)

MadsTorgersen replied:
Good catch; fixed in next commit

KalleOlaviNiemitalo commented on standard/expressions.md:
```diff
@@ -5984,6 +6058,27 @@ When a property or indexer declared in a *struct_type* is the target of an assig
+- Then the right hand side `E` is deconstructed into a list of expressions `E1, E2, ...` as follows:
+  - If `E` is a tuple expression with the same arity as the left-hand side, then `E1, E2, ...` shall be the element expressions of `E`.
+  - Otherwise, if `E` has a tuple type (§tuple-types-new-clause) with the same arity as the left-hand side, then `E1, E2, ...` shall be the member access expressions `E.Item1, E.Item2, ...`, except that `E` shall be evaluated only once.
+  - Otherwise, if `E.Deconstruct(out var v1, out var v2, ...)` with the number of arguments corresponding to the arity of the left hand side is a valid instance method invocation, then that invocation shall be performed and `E1, E2, ...` shall be the simple name expressions `v1, v2, ...`.
```
(No changes requested)

I was surprised that `v1` etc. in the Deconstruct call always use `var`, rather than the type of the tuple element at the left side when known. This prevents one from using the types to disambiguate the call when Deconstruct is overloaded. On the other hand, I suppose this limitation is what makes it possible to allow implicit conversions from the types of the parameters to the types of the tuple elements.

```
static class C {
    static void M() {
        int a, b;
        
        (a, b) = default(D); // error, ambiguous
        default(D).Deconstruct(out a, out b); // OK
        
        (a, b) = default(E); // OK, implicit conversion from short to int
    }
}

struct D {
    public void Deconstruct(out int a, out int b) {
        a = b = 0;
    }
    public void Deconstruct(out string a, out string b) {
        a = b = "";
    }
}

struct E {
    public void Deconstruct(out short a, out short b) {
        a = b = 0;
    }
}
```

MadsTorgersen replied:
Correct, we don't try to get smart about overloaded Deconstructors of the same arity.

RexJaeschke commented:
Re discards, I propose adding the following text to lexical-structure.md, near the end, just before the mention of identifiers starting with `__`:

> The semantics of an identifier named `_` depends on the context in which it appears:
> - It can denote a named program element, such as a variable, class, or method, or
> - It can denote a discard (§discards-new-clause)

RexJaeschke commented:
Some editorial feedback:

1. Replace all "legal" with "valid."
2. For grammar blocks, replace all "[backticks][space]ANTLR" with "[backticks]ANTLR", so the tools can find them.
3. For C# code blocks, replace "[backticks][space]c#" with "[backticks]csharp", so the tools can find them.

MadsTorgersen committed:
Add examples

MadsTorgersen committed:
Fix editorial comments

MadsTorgersen commented:
> Re discards, I propose adding the following text to lexical-structure.md, near the end, just before the mention of identifiers starting with `__`:
> 
> > The semantics of an identifier named `_` depends on the context in which it appears:
> > 
> > * It can denote a named program element, such as a variable, class, or method, or
> > * It can denote a discard (§discards-new-clause)

Done.

MadsTorgersen commented:
> Some editorial feedback:
> 
> 1. Replace all "legal" with "valid."
> 2. For grammar blocks, replace all "[backticks][space]ANTLR" with "[backticks]ANTLR", so the tools can find them.
> 3. For C# code blocks, replace "[backticks][space]c#" with "[backticks]csharp", so the tools can find them.

Done. Thanks!

Nigel-Ecma commented on standard/conversions.md:
```diff
@@ -310,6 +315,24 @@ The following further implicit conversions exist for a given type parameter `T`
 
+### §implicit-tuple-conversions-new-clause Implicit tuple conversions
+
+An implicit conversion exists from a tuple expression `E` to a tuple type `T` if `E` has the same arity as `T` and an implicit conversion exists from each element in `E` to the corresponding element type in `T`. The conversion is performed by creating an instance of `T`'s corresponding `System.ValueTuple<...>` type, and initializing each of its fields in order by evaluating the corresponding tupl ...
```
There are five introduced "in order" phrases, only one has "left to right" specified. No exhaustive search done but *§11.6.2.3 Run-time evaluation of argument lists* does specify the order (left to right unsurprisingly!). Should we consistently do so?

RexJaeschke replied:
TG2 call of 2023-02-01: Let's explicitly say the order for all phrases.

MadsTorgersen replied:
Fixed all occurrences that are related to this PR in upcoming commit

Nigel-Ecma commented on standard/expressions.md:
```diff
@@ -37,6 +38,7 @@ Most of the constructs that involve an expression ultimately require the express
 - The value of a variable is simply the value currently stored in the storage location identified by the variable. A variable shall be considered definitely assigned ([§9.4](variables.md#94-definite-assignment)) before its value can be obtained, or otherwise a compile-time error occurs.
 - The value of a property access expression is obtained by invoking the *get_accessor* of the property. If the property has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed, and the result of the invocation becomes the value of the property access expression.
 - The value of an indexer access expression is obtained by invoking the *get_accessor* of the indexer. If the indexer has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed with the argument list associated with the indexer access expression, and the result of the invocation becomes the v ...
+- The value of a tuple expression is obtained by applying an implicit tuple conversion (§implicit-tuple-conversions-new-clause) to the type of the tuple expression itself. It is an error to obtain the value of a tuple expression that does not have a type.
```
The first sentence doesn't read well to me. Further §tuple-expressions-new-clause appears to list three (or maybe just two as one leads here and another to §implicit-tuple-conversions-new-clause which is where this one leads...) ways to obtain a tuple value from a tuple expression, so not quite "is obtained" but "can be obtained"?

jskeet replied:
While I think this should be discussed, I *think* it's okay, because of the context in which this clause is used. And I don't think it could be "can be obtained" because this needs to specify what value *is* obtained when there isn't an explicit conversion involved, for example. I could well be wrong though - not thinking terribly clearly today.

MadsTorgersen replied:
After discussion, let's remove "itself"

MadsTorgersen replied:
Fixed in upcoming commit

MadsTorgersen committed:
Fix declaration spaces and simple name lookup

MadsTorgersen committed:
Update basic-concepts.md

MadsTorgersen committed:
Fix local declarations

MadsTorgersen committed:
Update statements.md

MadsTorgersen commented:
Current status:
- I found that my understanding of the behavior of deconstruction did not fully match the compiler behavior. I am still chasing this down, so the section on deconstructing assignments remains incomplete. It still needs:
  - Correct semantics
  - Examples
  - A description of the `var (x, y)` syntax
- I need to investigate whether changes are needed to overload resolution in the presence of out variables.
- I need to investigate whether definite assignment is affected by declaration expressions.

Nigel-Ecma reviewed (commented):
Just a couple of comments. Looking good so far.

jskeet commented on standard/basic-concepts.md:
```diff
@@ -78,12 +78,16 @@ There are several different types of declaration spaces, as described in the fol
 - Within all compilation units of a program, *namespace_member_declaration*s within *namespace_declaration*s that have the same fully qualified namespace name are members of a single combined declaration space.
 - Each *compilation_unit* and *namespace_body* has an ***alias declaration space***. Each *extern_alias_directive* and *using_alias_directive* of the *compilation_unit* or *namespace_body* contributes a member to the alias declaration space ([§13.5.2](namespaces.md#1352-using-alias-directives)).
-- Each non-partial class, struct, or interface declaration creates a new declaration space. Each partial class, struct, or interface declaration contributes to a declaration space shared by all matching parts in the same program ([§15.2.3](structs.md#1523-partial-modifier)).Names are introduced into this declaration space through *class_member_declaration*s, *struct_member_declaration*s, *interfa ...
+- Each non-partial class, struct, or interface declaration creates a new declaration space. Each partial class, struct, or interface declaration contributes to a declaration space shared by all matching parts in the same program ([§15.2.3](structs.md#1523-partial-modifier)). Names are introduced into this declaration space through *class_member_declaration*s, *struct_member_declaration*s, *interf ...
```
For those trying to find a diff here - there's now a space before "Names" at the start of the third sentence :)

jskeet commented on standard/basic-concepts.md:
```diff
@@ -78,12 +78,16 @@ There are several different types of declaration spaces, as described in the fol
-- Each *block* or *switch_block*, as well as a `for`, `foreach`, and `using` statement, creates a local variable declaration space for local variables and local constants. Names are introduced into this declaration space through *local_variable_declaration*s and *local_constant_declaration*s. Note that blocks that occur as or within the body of a function member, anonymous function, or local func ...
-
+- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter ...
+- Additional local variable declaration spaces may occur within member declarations, anonymous functions and local functions. Names are introduced into these declaration spaces through *declaration_expression*s and *declaration_statement*s. Local variable declaration spaces may be nested, but it is an error for a local variable declaration space and a nested local variable declaration space to co ...
```
I wonder whether "declare" is clearer than "contain" here?

```suggestion
- Additional local variable declaration spaces may occur within member declarations, anonymous functions and local functions. Names are introduced into these declaration spaces through *declaration_expression*s and *declaration_statement*s. Local variable declaration spaces may be nested, but it is an error for a local variable declaration space and a nested local variable declaration space to declare elements with the same name. Thus, within a nested declaration space it is not possible to declare a local variable or constant with the same name as a parameter, type parameter, local variable or constant in an enclosing declaration space. It is possible for two declaration spaces to declare elements with the same name as long as neither declaration space contains the other. Local declaration spaces are created by the following constructs:
```

RexJaeschke replied:
TG2 call of 2023-02-01: Jon and Mads will look at this offline.

BillWagner replied:
I think "contains" is the correct verb in this instance. I don't have strong feelings though.

MadsTorgersen replied:
This is the current terminology in use. Feel free to raise an issue to change.

jskeet commented on standard/basic-concepts.md:
```diff
@@ -78,12 +78,16 @@ There are several different types of declaration spaces, as described in the fol
-
+- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter ...
+- Additional local variable declaration spaces may occur within member declarations, anonymous functions and local functions. Names are introduced into these declaration spaces through *declaration_expression*s and *declaration_statement*s. Local variable declaration spaces may be nested, but it is an error for a local variable declaration space and a nested local variable declaration space to co ...
+  - Each *variable_initializer* introduces its own local variable declaration space, that is not nested within any other local variable declaration space.
```
Are variable_initializers only for fields, or do they include local variable initializers too?

RexJaeschke replied:
TG2 call of 2023-02-01, Mads will check offline: fix or clarify

BillWagner replied:
Variable initializers only apply to fields.  See [§14.5.6](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/classes.md#1456-variable-initializers) Variable initializers

MadsTorgersen replied:
They apply to properties as well (See https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/classes.md#1471-general), but nowhere else.

MadsTorgersen replied:
Clarified in upcoming commit.

jskeet commented on standard/conversions.md:
```diff
@@ -78,7 +79,11 @@ However, dynamic conversions ([§10.2.10](conversions.md#10210-implicit-dynamic-
 An identity conversion converts from any type to the same type. One reason this conversion exists is so that a type T or an expression of type T can be said to be convertible to T itself.
 
-Because `object` and `dynamic` are considered equivalent there is an identity conversion between `object` and `dynamic`, and between constructed types that are the same when replacing all occurrences of `dynamic` with `object`.
+In some cases there is an identity conversion between types that are not exactly the same, but are considered equivalent. Such identity conversions exist:
```
I think we need to be a little more descriptive than "considered equivalent" - we're not really saying that `object` and `dynamic` are considered equivalent *in general* (there are marked differences). Considered equivalent at runtime?

BillWagner replied:
The same language is used in [§10.2.2](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/conversions.md#1022-identity-conversion) Identity conversion

See #730

MadsTorgersen replied:
I'm resolving it here since this PR isn't introducing the verbiage, and a separate issue has been opened.

jskeet commented on standard/expressions.md:
```diff
@@ -1446,7 +1448,7 @@ simple_name
 A *simple_name* is either of the form `I` or of the form `I<A₁, ..., Aₑ>`, where `I` is a single identifier and `I<A₁, ..., Aₑ>` is an optional *type_argument_list*. When no *type_argument_list* is specified, consider `e` to be zero. The *simple_name* is evaluated and classified as follows:
 
-- If `e` is zero and the *simple_name* appears within a *block* and if the *block*’s (or an enclosing *block*’s) local variable declaration space ([§7.3](basic-concepts.md#73-declarations)) contains a local variable, parameter or constant with name `I`, then the *simple_name* refers to that local variable, parameter or constant and is classified as a variable or value.
+- If `e` is zero and the *simple_name* appears within a local variable declaration space ([§7.3](basic-concepts.md#73-declarations)) that contains a local variable, parameter or constant with name `I`, then the *simple_name* refers to that local variable, parameter or constant and is classified as a variable or value.
```
Again I think I might be happier with declares rather than contains. I think it's fine to say "this block declares variable x" or "this block contains a local variable declaration for x" - but "this block contains a local variable x" feels off to me. (I realize this isn't new language.)

MadsTorgersen replied:
Discussion: change to "directly contains"

MadsTorgersen replied:
Fixed in upcoming commit

jskeet commented on standard/expressions.md:
```diff
@@ -1480,6 +1487,34 @@ parenthesized_expression
+(int i, string) t1 = (i: 1, "One");
+(int i, string) t2 = (i: 2, null);
+var t3 = (i: 3, "Three"); // (int i, string)
+var t4 = (i: 4, null); // Error: no type
```
Would it be worth including an example where we end up with "necessary but odd-looking" parentheses, e.g.

```csharp
SomeMethod((1, 2));
```

(Obviously with more description!) It's the kind of thing I'd definitely mention in a tutorial - and I know the standard isn't a tutorial, but if we're already giving examples, perhaps one would be useful?

MadsTorgersen replied:
Example added in upcoming commit

jskeet commented on standard/expressions.md:
````diff
@@ -1480,6 +1487,34 @@ parenthesized_expression
+var t4 = (i: 4, null); // Error: no type
+```
+
+In this example, all four tuple expressions are valid. The first two, `t1` and `t2`, do not use the type of the tuple expression, but instead apply an implicit tuple conversion. The third tuple expression has a type `(int i, string)`, and can therefore be reclassified as a value of that type. The declaration of `t4`, on the other hand, is an error: The tuple expression has no type because its sec ...
````
Might it be useful to change the declaration of `t2` to use `(long x, string)` and note the implicit int to long conversion?

MadsTorgersen replied:
Good idea. Fixed in upcoming commit.

jskeet commented on standard/expressions.md:
````diff
@@ -4339,6 +4374,54 @@ A *throw expression* shall only occur in the following syntactic contexts:
+(int i1, int _, (var i2, var _), _) = (1, 2, (3, 4), 5);
+```
+
+This example shows the use of implicitly and explicitly typed declaration expressions for both variables and discards in a deconstructing assignment.
````
If a discard with an explicit type is used on the LHS, and it needs an explicit conversion from the type on the RHS, presumably the explicit conversion is still performed (potentially with side-effects such as an exception) even if the result is going to be discarded, right?

MadsTorgersen replied:
Yes.

jskeet commented on standard/expressions.md:
```diff
@@ -5984,6 +6069,27 @@ When a property or indexer declared in a *struct_type* is the target of an assig
+
+If the left operand of a `=` operator is classified as a tuple, the right hand side is ***deconstructed*** into individual expressions, which are assigned to each of the elements of the left hand side tuple.
+
+If any of the tuple elements of the left operand have an identifier, a compile-time error occurs. If any of the tuple elements of the left operand is a declaration expression, and any other element is not a declaration expression or discard, a compile-time error occurs.
```
```suggestion
If any of the tuple elements of the left operand has an identifier, a compile-time error occurs. If any of the tuple elements of the left operand is a declaration expression, and any other element is not a declaration expression or discard, a compile-time error occurs.
```

(Just to be consistent in terms of singular or plural.)

jskeet replied:
I think an example showing both of these would be useful though. In my brain-fogged state, I don't fully understand the first constraint.

Or possibly I don't understand why tuple elements are used. Where do we prevent this from being valid, for example:

```csharp
(1, 2) = (3, 4);
```

MadsTorgersen replied:
The singular/plural is fixed in upcoming commit. As for legality of the LHS, this is clarified in the new verbiage, in a paragraph that (recursively if necessary) either finds the type of a simple assignment or deems it invalid.

jskeet commented on standard/expressions.md:
```diff
@@ -5984,6 +6069,27 @@ When a property or indexer declared in a *struct_type* is the target of an assig
+
+If any of the tuple elements of the left operand have an identifier, a compile-time error occurs. If any of the tuple elements of the left operand is a declaration expression, and any other element is not a declaration expression or discard, a compile-time error occurs.
+
+The evaluation of a deconstructing assignment `(R1, ..., Rn) = E` proceeds as follows:
```
Is there a reason for picking R here? Could we use L instead, given that it's on the left side of the operator?

MadsTorgersen replied:
Changed naming in the new formulation - they are now `x1...xn` and `y1...yn` respectively, to match existing naming in the simple assignment section.

jskeet commented on standard/statements.md:
```diff
@@ -278,18 +278,20 @@ In addition to the reachability provided by normal flow of control, a labeled st
 ### 12.6.1 General
 
-A *declaration_statement* declares a local variable, local constant, or local function. Declaration statements are permitted in blocks, but are not permitted as embedded statements.
+A *declaration_statement* declares one or more local variables, a local constant, or a local function. Declaration statements are permitted in blocks, but are not permitted as embedded statements.
```
```suggestion
A *declaration_statement* declares one or more local variables, one or more local constants, or a local function. Declaration statements are permitted in blocks, but are not permitted as embedded statements.
```

(I believe `const int x = 5, y = 10;` declares two local constants.)

MadsTorgersen replied:
Fixed in upcoming commit

jskeet commented on standard/variables.md:
```diff
@@ -150,6 +150,23 @@ A local variable introduced by a *local_variable_declaration* is not automatical
+
+The assignment to `_` shows a simple pattern for ignoring the result of an expression.
+The call of `M` shows the different forms of discards available in tuples and as out parameters.
+
```
We need an end example :)

MadsTorgersen replied:
All examples formatted in upcoming commit

jskeet reviewed (commented):
Various comments. I'm not thinking straight enough to be ready to say "this looks good" or "we're missing XYZ" - but I'm confident it's in the right direction, and I don't *know* of anything we're missing...

BillWagner commented on standard/variables.md:
```diff
@@ -150,6 +150,23 @@ A local variable introduced by a *local_variable_declaration* is not automatical
 
+#### §discards-new-clause Discards
+
+A ***discard*** is a local variable that has no name and can thus only be referred to once, at the point where it is introduced. A discard is not initially assigned, so it is always an error to access its value. A discard is introduced by a *declaration_expression* with the identifier `_`, and by use of the *simple_name* `_` when lookup of that name does not find a declaration ([§11.7.4](expressi ...
```
Meeting suggestion: Add a *note* to point out that `_` is a valid identifier, and is a ***discard*** only when lookup doesn't find a variable `_`.

MadsTorgersen replied:
Note added in upcoming commit

MadsTorgersen commented on standard/expressions.md:
```diff
@@ -1466,6 +1468,11 @@ A *simple_name* is either of the form `I` or of the form `I<A₁, ..., Aₑ>`,
     - Otherwise, if the namespaces imported by the *using_namespace_directive*s of the namespace declaration contain exactly one type having name `I` and `e` type parameters, then the *simple_name* refers to that type constructed with the given type arguments.
     - Otherwise, if the namespaces imported by the *using_namespace_directive*s of the namespace declaration contain more than one type having name `I` and `e` type parameters, then the *simple_name* is ambiguous and a compile-time error occurs.  
   > *Note*: This entire step is exactly parallel to the corresponding step in the processing of a *namespace_or_type_name* ([§7.8](basic-concepts.md#78-namespace-and-type-names)). *end note*
+- Otherwise, if `e` is zero and `I` is the identifier `_`, the *simple_name* is a fresh discard (§discards-new-clause) and its type shall be inferred from the syntactic context:
```
Discussion: remove "fresh"

MadsTorgersen replied:
Fixed in upcoming commit

BillWagner commented on standard/expressions.md:
```diff
@@ -1480,6 +1487,34 @@ parenthesized_expression
 
 A *parenthesized_expression* is evaluated by evaluating the *expression* within the parentheses. If the *expression* within the parentheses denotes a namespace or type, a compile-time error occurs. Otherwise, the result of the *parenthesized_expression* is the result of the evaluation of the contained *expression*.
 
+### §tuple-expressions-new-clause Tuple expressions
```
Tuples depend on `ValueTuple` with all the arities.  That should be added to the Annex on standard types.

MadsTorgersen replied:
I've added them all in the upcoming commit. There is still no explanation for how longer tuples are chained.

MadsTorgersen commented on standard/expressions.md:
```diff
@@ -4339,6 +4374,54 @@ A *throw expression* shall only occur in the following syntactic contexts:
+- As an `out` *argument_value* in an *argument_list*.
+- As a *tuple_element* in a *tuple_expression* that occurs on the left-hand side of a deconstructing assignment.
+
+It is an error for a variable declared with a *declaration_expression* to be referenced within the *argument_list* or left-hand side of a deconstructing assignment where it occurs.
```
Should exclude the whole deconstructing assignment

MadsTorgersen replied:
Fixed.

BillWagner commented on standard/expressions.md:
```diff
@@ -4339,6 +4374,54 @@ A *throw expression* shall only occur in the following syntactic contexts:
 - As the second operand of a null coalescing operator (`??`).
 - As the body of an expression-bodied lambda or member.
 
+## §declaration-expressions-new-clause Declaration expressions
```
We may want to make a note that this is legal:

```
public class C {
    public void M(out int i) {
    }
    
    public void M2()
    {
        String _ = "bad";
        M(out int _);
}
```

The point being that the first `_` is a variable named `_`, and the declaration expression introduces a variable, but does not introduce a name.

MadsTorgersen replied:
Example added in upcoming commit

BillWagner commented on standard/variables.md:
```diff
@@ -113,9 +113,9 @@ Within an instance constructor of a struct type, the `this` keyword behaves exac
+A ***local variable*** is declared by a *local_variable_declaration*, *declaration_expression*, *foreach_statement*, or *specific_catch_clause* of a *try_statement*. For a *foreach_statement*, the local variable is an iteration variable ([§12.9.5](statements.md#1295-the-foreach-statement)). For a *specific_catch_clause*, the local variable is an exception variable ([§12.11](statements.md#1211-the ...
 
-A *local_variable_declaration* can occur in a *block*, a *for_statement*, a *switch_block*, or a *using_statement*.
+A *local_variable_declaration* can occur in a *block*, a *for_statement*, a *switch_block*, or a *using_statement*. A *declaration_expression* can occur as an `out` *argument_value*, and in a *tuple_element* that is the target of a deconstructing assignment (§deconstructing-assignment-new-clause).
```
nit: I might consider "as" in both clauses:

```suggestion
A *local_variable_declaration* can occur in a *block*, a *for_statement*, a *switch_block*, or a *using_statement*. A *declaration_expression* can occur as an `out` *argument_value*, and as a *tuple_element* that is the target of a deconstructing assignment (§deconstructing-assignment-new-clause).
```

MadsTorgersen replied:
Fixed in upcoming commit

MadsTorgersen commented on standard/variables.md:
```diff
@@ -150,6 +150,23 @@ A local variable introduced by a *local_variable_declaration* is not automatical
 
+#### §discards-new-clause Discards
+
+A ***discard*** is a local variable that has no name and can thus only be referred to once, at the point where it is introduced. A discard is not initially assigned, so it is always an error to access its value. A discard is introduced by a *declaration_expression* with the identifier `_`, and by use of the *simple_name* `_` when lookup of that name does not find a declaration ([§11.7.4](expressi ...
```
Discussion: This is a bit imprecise. Think of better wording?

MadsTorgersen replied:
Tried to reword more precisely in upcoming commit

KalleOlaviNiemitalo commented on standard/expressions.md:
```diff
@@ -4339,6 +4374,54 @@ A *throw expression* shall only occur in the following syntactic contexts:
+- As an `out` *argument_value* in an *argument_list*.
+- As a *tuple_element* in a *tuple_expression* that occurs on the left-hand side of a deconstructing assignment.
+
+It is an error for a variable declared with a *declaration_expression* to be referenced within the *argument_list* or left-hand side of a deconstructing assignment where it occurs.
```
Should both of the following `N` calls be errors, then? SharpLab accepts the first one but I'm not sure which version of C# it uses.

```csharp
public class C {
    public void M() {
        N(out int i, out i); // OK
        N(out var j, out j); // error CS8196
    }
    void N(out int a, out int b) {
        a = b = 0;
    }
}
```

MadsTorgersen replied:
You are right, the rule I am expressing is too strict. The first call to N should be allowed in your example. Fixed in upcoming commit.

MadsTorgersen committed:
Fix deconstructing assignment

Deconstructing assignment is now fully integrated into simple assignment, and doesn't have its own section.

Deconstruction itself has been broken into an independent subsection of Expressions in anticipation of more deconstruction to come in future releases.

The evaluation order of a deconstructing assignment is now elegant and recursive. However, it still does not match the compiler behavior 100%. The evaluation order of the expressions directly occurring in the assignment expression is correct (except for what seems to be a compiler bug in a very specific case). However, in the case of nested left side tuples, some conversions and deconstructions may occur slightly later than my text implies. I've chosen to live with this difference since it is small, unobservable when the conversions and deconstructions are side-effect free (as they should be), and frankly beyond my capabilities to adequately overcome.

MadsTorgersen committed:
Fix linting issues

MadsTorgersen committed:
Fix more linting issues

MadsTorgersen committed:
Fix issues from comments

MadsTorgersen committed:
Try to fix linting issue

MadsTorgersen committed:
Fix linting issue

MadsTorgersen commented:
Current status:

Deconstructing assignment is now functionally complete, modulo unknown issues.

The remaining work that I am aware of is:

- Add examples of deconstructing assignment
- Annotate examples for testing throughout
- A description of the `var (x, y)` syntax
- Investigate whether changes are needed to overload resolution in the presence of out variables.
- Investigate whether definite assignment is affected by declaration expressions.

gafter commented:
Reviewing this now.

gafter commented on standard/basic-concepts.md:
```diff
@@ -78,12 +78,16 @@ There are several different types of declaration spaces, as described in the fol
-- Each *block* or *switch_block*, as well as a `for`, `foreach`, and `using` statement, creates a local variable declaration space for local variables and local constants. Names are introduced into this declaration space through *local_variable_declaration*s and *local_constant_declaration*s. Note that blocks that occur as or within the body of a function member, anonymous function, or local func ...
-
+- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter ...
+- Additional local variable declaration spaces may occur within member declarations, anonymous functions and local functions. Names are introduced into these declaration spaces through *declaration_expression*s and *declaration_statement*s. Local variable declaration spaces may be nested, but it is an error for a local variable declaration space and a nested local variable declaration space to co ...
```
> Additional local variable declaration spaces **may** occur

What do you mean "**may**"?  Do they or don't they?  Does it depend on the day of the week 😉?

You probably want to say something about pattern variables introducing locals as well.

MadsTorgersen replied:
"*may*": Doesn't the subsequent "Local declaration spaces are created by the following constructs:" address that?

*Pattern variables*: I left everything specifically pertaining to patterns (and their effect on switch blocks) out of this PR. We should discuss how it's best brought in. I'd imagined the pattern matching PR could add those bits, but I don't really have a preference.

gafter replied:
Unfortunately, there is no place to put those bits because this isn't merged yet.

BillWagner replied:
Proposal for this meeting: Let's make an issue to address the pattern variables introducing locals. Then, we can add it once both prerequisite PRs are merged.

MadsTorgersen replied:
Let's fix this when at least one of #664 and #757 is merged.

gafter commented on standard/basic-concepts.md:
```diff
@@ -78,12 +78,16 @@ There are several different types of declaration spaces, as described in the fol
+- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter ...
+- Additional local variable declaration spaces may occur within member declarations, anonymous functions and local functions. Names are introduced into these declaration spaces through *declaration_expression*s and *declaration_statement*s. Local variable declaration spaces may be nested, but it is an error for a local variable declaration space and a nested local variable declaration space to co ...
+  - Each *variable_initializer* in a field and property declaration introduces its own local variable declaration space, that is not nested within any other local variable declaration space.
+  - The body of a function member, anonymous function, or local function, if any, creates a local variable declaration space that is considered to be nested within the function's local variable declaration space.
```
It isn't clear if this includes indexer accessor expression bodies (I am not sure if they are function member bodies).

MadsTorgersen replied:
They are: [11.6.1 General](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1161-general).

gafter commented on standard/basic-concepts.md:
```diff
@@ -78,12 +78,16 @@ There are several different types of declaration spaces, as described in the fol
+  - The body of a function member, anonymous function, or local function, if any, creates a local variable declaration space that is considered to be nested within the function's local variable declaration space.
+  - Each *constructor_initializer* creates a local variable declaration space nested within the instance constructor declaration. The local variable declaration space for the constructor body is in turn nested within this local variable declaration space.
+  - Each *block*, *switch_block*, *iteration_statement* and *using_statement* creates a nested local variable declaration space.
+  - Each *embedded_statement* that is not directly part of a *statement_list* creates a nested local variable declaration space.
```
Additional contexts that might not be covered by this yet:
- Pattern variables declared in a switch (case) label.  This one is subtle because it introduces a scope that encompasses just the switch section, nested within the switch block, but local variables declared in the statements of the switch section go in the scope of the switch block.
- A switch case guard (`when` clause).  [These go in just the switch block's scope](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGZwBsmALOgLIAUA8gEYBWApgMYAu6rAlIcQJJiI9AGcA7gEtO7ABbpaffANGj2AQzGt0cdBLmsAdjVoB7AK7cAbhoBO6AB68Qqte8Z3WGgNYBuN3d0TW10GH1DEzpLG3snF0Cg4k9vf0SiAF83LOFiKDJ0RjMzABtTGPQpI25nfCd0AF5dP0wAdnROOwtWFpycoA) (see previous bullet).
- [A query clause](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.3/expression-variables-in-initializers.md)
- A catch block, where the exception variable is introduced.  Its absence looks like a bug in the previous version.  This will take care of expression variables in an exception filter as well.

MadsTorgersen replied:
*Pattern variables*: I left those out for now. Let's discuss which PR should do them.
*Switch guards*: Same.
*Query clauses*: We've generally taken the stance that query clauses are defined by their desugaring. 
*Catch blocks*: Good catch! It's orthogonal to the subject of this PR. I can add it in anyway, or we can open an issue. What do you think?

gafter replied:
Query clauses define most of their their semantics by the translation, but the scoping is not based on that translation.  These scopes only came into existence in C# 7 because of its expression variables.  [Evidence here](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lqZwB0AMgJYB2AjvkgQMyYzoDC6A3suj5k1NnQBZABQAjAPYSANujEBKTt14qAbgEMATulXoAvMpVH0AM00SAtugAe6KsLgj5h47wDuACwCmmr8JgiEgCuAC46WujUzkiuRgDOXtJeAMZhQgHBYVF0KgC+LlD8cABsRAwAPADKIZpUAOYAfA5OSjGuUADs6JRB0tIAhDm8+W08hUSlkjL+gaFyUrJRrbHU+ug1QV5Dxp3rmpvb6CMjQA===).

I think it is something like this:
- Variables introduced in the expression of a `from` clause are in scope for the scope containing the query expression.
- For other clauses, each expression of the query introduces its own scope that contains only that expression.

MadsTorgersen replied:
Captured in #768 to be addressed post-merge.

gafter commented on standard/expressions.md:
```diff
@@ -14,6 +14,7 @@ The result of an expression is classified as one of the following:
 - A variable. Every variable has an associated type, namely the declared type of the variable.
 - A null literal. An expression with this classification can be implicitly converted to a reference type or nullable value type.
 - An anonymous function. An expression with this classification can be implicitly converted to a compatible delegate type or expression tree type.
+- A tuple. Every tuple has a fixed number of elements, each with an expression and an optional tuple element name.
```
I suspect the words "The result of" should be removed from the introductory sentence:

> The result of an expression is classified as one of the following:

I think this list is intended to contain all of the expression forms that have target-type semantics.  But a number of recently-introduced target-typed forms are not listed.  For example, what about interpolated strings?

MadsTorgersen replied:
This is orthogonal to this PR. Would you mind opening an issue?

gafter commented on standard/expressions.md:
```diff
@@ -37,6 +38,7 @@ Most of the constructs that involve an expression ultimately require the express
 - The value of a variable is simply the value currently stored in the storage location identified by the variable. A variable shall be considered definitely assigned ([§9.4](variables.md#94-definite-assignment)) before its value can be obtained, or otherwise a compile-time error occurs.
 - The value of a property access expression is obtained by invoking the *get_accessor* of the property. If the property has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed, and the result of the invocation becomes the value of the property access expression.
 - The value of an indexer access expression is obtained by invoking the *get_accessor* of the indexer. If the indexer has no *get_accessor*, a compile-time error occurs. Otherwise, a function member invocation ([§11.6.6](expressions.md#1166-function-member-invocation)) is performed with the argument list associated with the indexer access expression, and the result of the invocation becomes the v ...
+- The value of a tuple expression is obtained by applying an implicit tuple conversion (§implicit-tuple-conversions-new-clause) to the type of the tuple expression. It is an error to obtain the value of a tuple expression that does not have a type.
```
No sure exactly what "obtain the value of" means.

MadsTorgersen replied:
This is the same exact phrasing as the existing cases.

gafter commented on standard/expressions.md:
````diff
@@ -1480,6 +1501,42 @@ parenthesized_expression
+    : (identifier ':')? expression
+```
+
+A tuple expression is classified as a tuple. It has a type if all the element expressions have a type. In that case, the type of a tuple expression is the tuple type of the same arity, where each type element has the type - and name, where given - of the corresponding element expression.
````
This doesn't take into account [inferred tuple names](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.1/infer-tuple-names.md).

MadsTorgersen replied:
I forgot about those! Fixed in upcoming commit.

gafter commented on standard/expressions.md:
```diff
@@ -4339,6 +4396,68 @@ A *throw expression* shall only occur in the following syntactic contexts:
+- As an `out` *argument_value* in an *argument_list*.
+- As a *tuple_element* in a *tuple_expression* that occurs on the left side of a deconstructing assignment ([§11.19.2](expressions.md#11192-simple-assignment)).
+
+It is an error for an implicitly typed variable declared with a *declaration_expression* to be referenced within the *argument_list* where it occurs.
```
```suggestion
It is an error for an implicitly typed variable declared with a *declaration_expression* to be referenced within the *argument_list* where it is declared.
```

MadsTorgersen replied:
Fixed in upcoming PR

gafter commented on standard/expressions.md:
```diff
@@ -5863,23 +5982,31 @@ The `=` operator is called the simple assignment operator.
 If the left operand of a simple assignment is of the form `E.P` or `E[Ei]` where `E` has the compile-time type `dynamic`, then the assignment is dynamically bound ([§11.3.3](expressions.md#1133-dynamic-binding)). In this case, the compile-time type of the assignment expression is `dynamic`, and the resolution described below will take place at run-time based on the run-time type of `E`. If the le ...
 
-In a simple assignment, the right operand shall be an expression that is implicitly convertible to the type of the left operand. The operation assigns the value of the right operand to the variable, property, or indexer element given by the left operand.
+A simple assignment with a tuple as the left operand is also called a ***deconstructing assignment***. If any of the tuple elements of the left operand has an identifier, a compile-time error occurs. If any of the tuple elements of the left operand is a declaration expression, and any other element is not a declaration expression or discard, a compile-time error occurs.
```
```
(int x, int y) a = (1, 2), b = (3, 4);
a = b; // is this a deconstructing assignment? Is it an error because of the names?
```

I don't really understand "If any of the tuple elements of the left operand has an identifier, a compile-time error occurs."

gafter replied:
```
(int x, int y) a = (1, 2);
(int x, int y) b = a; // Is it an error because of the names?
```

MadsTorgersen replied:
The "with a tuple as the left operand" phrasing above is unhelpfully imprecise. It should say "where the left operand is classified as a tuple" or something like that. In other words it needs to be a tuple _expression_, and not just something of tuple type in order for the assignment to be a deconstructing assignment.

MadsTorgersen replied:
Fixed in upcoming commit

gafter commented on standard/statements.md:
```diff
@@ -278,18 +278,20 @@ In addition to the reachability provided by normal flow of control, a labeled st
 
 A local variable is declared using a *local_variable_declaration* ([§12.6.2](statements.md#1262-local-variable-declarations)). A local constant is declared using a *local_constant_declaration* ([§12.6.3](statements.md#1263-local-constant-declarations)). A local function is declared using a *local_function_declaration* ([§12.6.4](statements.md#1264-local-function-declarations)).
 
+The declared names are introduced into the nearest enclosing declaration space ([§7.3](basic-concepts.md#73-declarations)).
```
There is an exception for this in a switch block, as discussed earlier.

MadsTorgersen replied:
Yes. I didn't add anything pertaining to the expanded switch statements.

MadsTorgersen replied:
Punting everything pertaining to pattern matching out of this PR

gafter commented on standard/statements.md:
```diff
@@ -278,18 +278,20 @@ In addition to the reachability provided by normal flow of control, a labeled st
 ### 12.6.1 General
 
-A *declaration_statement* declares a local variable, local constant, or local function. Declaration statements are permitted in blocks, but are not permitted as embedded statements.
+A *declaration_statement* declares one or more local variables, one or more local constants, or a local function. Declaration statements are permitted in blocks, but are not permitted as embedded statements.
```
Are they permitted in switch blocks too?

MadsTorgersen replied:
I believe so. Seems like a pre-existing omission.

MadsTorgersen replied:
Fixed in upcoming commit

gafter commented on standard/standard-library.md:
```diff
@@ -245,6 +245,75 @@ namespace System
     public struct UInt64 { }
     public struct UIntPtr { }
 
+    public struct ValueTuple<T1>
```
There is also a zero-arity `ValueTuple`.

MadsTorgersen replied:
There is, but I don't think the language depends on it. Am I forgetting something?

gafter replied:
I don't think so.

gafter commented on standard/types.md:
```diff
@@ -378,6 +387,40 @@ No standard conversions exist between `bool` and other value types. In particula
 
+### §tuple-types-new-clause Tuple types
+
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...T ...
```
> which shall be a set of generic struct types capable of expressing every arity of tuple types

That isn't the way long tuple types are constructed.  `System.ValueTuple<int, int, int, int, int, int, int, int, int>` is an error.

MadsTorgersen replied:
I picked weasel words to not directly say *how* they are capable of representing long tuples. But based on feedback from last meeting, we probably need to say a bit more about the scheme. I'll try to think of a balanced way to add more detail but not too much.

gafter replied:
When you say "every arity", I assume you mean every arity greater than 1.  Because these are errors:
```
        System.ValueTuple x;
        if (x == x) {}
        System.ValueTuple<int> y = new System.ValueTuple<int>(1);
        if (y == y) {}
```

MadsTorgersen replied:
Both issues fixed in upcoming commit.

gafter commented on standard/types.md:
```diff
@@ -378,6 +387,40 @@ No standard conversions exist between `bool` and other value types. In particula
+
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...T ...
+
+Element names within a tuple type shall be distinct. It is an error for an explicit tuple element name to be of the form `ItemX` where `X` is any sequence of non-`0`-initiated decimal digits that could represent the position of a tuple element, except as the name for that actual element.
```
You have not defined explicit tuple element names.

When saying "that actual element", *what* actual element are you referring to?

MadsTorgersen replied:
Thanks, I'll do some word smithing.

MadsTorgersen replied:
Fixed in upcoming commit.

gafter commented on standard/types.md:
```diff
@@ -378,6 +387,40 @@ No standard conversions exist between `bool` and other value types. In particula
+
+Element names within a tuple type shall be distinct. It is an error for an explicit tuple element name to be of the form `ItemX` where `X` is any sequence of non-`0`-initiated decimal digits that could represent the position of a tuple element, except as the name for that actual element.
+
+The optional element names are not represented in the `ValueTuple<...>` types, and are not stored in the runtime representation of a tuple value. There is an identity conversion between all tuple types with the same arity and the same sequence of element types, as well as to and from the corresponding constructed `ValueTuple<...>` type.
```
```suggestion
The optional element names are not represented in the `ValueTuple<...>` types, and are not stored in the runtime representation of a tuple value. There is an identity conversion between all tuple types with the same arity and identity-convertible sequences of element types, as well as to and from the corresponding constructed `ValueTuple<...>` type.
```

MadsTorgersen replied:
Fixed in upcoming commit.

gafter commented on standard/types.md:
```diff
@@ -378,6 +387,40 @@ No standard conversions exist between `bool` and other value types. In particula
+
+The optional element names are not represented in the `ValueTuple<...>` types, and are not stored in the runtime representation of a tuple value. There is an identity conversion between all tuple types with the same arity and the same sequence of element types, as well as to and from the corresponding constructed `ValueTuple<...>` type.
+
+The `new` operator [§11.7.15.2](expressions.md#117152-object-creation-expressions) cannot be applied directly to a tuple type. Tuple values can be created from tuple expressions (§tuple-expressions-new-clause), or by applying the `new` operator directly to a type constructed from `ValueTuple<...>`.
```
But you previously said that the tuple type syntax is a shorthand for an instantiation of `ValueTuple`.  I think you mean to say that the `new` operator cannot be used with the tuple type syntax.

MadsTorgersen replied:
Thanks, I'll change the wording.

MadsTorgersen replied:
Fixed in upcoming commit.

gafter commented:
One more thing.  I believe that deconstruction assignments are permitted to allow bidirectional type inference - from the left to the right and also from the right to the left - for different elements of the tuple.  That makes inferring the type of the whole assignment messy.  I don't know whether that belongs in this PR or in another tuple PR, but we should make sure we specify it.

gafter commented:
This is missing several special behaviors for tuples, e.g. the behavior of `==`.

BillWagner reviewed (commented):
I only had one additional note beyond what Neal has already identified.

MadsTorgersen committed:
Fix several comments from gafter

MadsTorgersen committed:
Fix "projection" tuple element names

MadsTorgersen committed:
Fix overload resolution with out vars

MadsTorgersen commented:
> One more thing. I believe that deconstruction assignments are permitted to allow bidirectional type inference - from the left to the right and also from the right to the left - for different elements of the tuple. That makes inferring the type of the whole assignment messy. I don't know whether that belongs in this PR or in another tuple PR, but we should make sure we specify it.

I believe this is adequately handled by the PR. The spec for simple assignment infers a type for the assignment as a whole, which, when the LHS is a tuple expression, takes the type of each element from the LHS if available, and the RHS otherwise.

jskeet commented on standard/types.md:
```diff
@@ -389,16 +389,21 @@ An enumeration type is a distinct type with named constants. Every enumeration t
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...T ...
 
-Element names within a tuple type shall be distinct. It is an error for an explicit tuple element name to be of the form `ItemX` where `X` is any sequence of non-`0`-initiated decimal digits that could represent the position of a tuple element, except as the name for that actual element.
+> *Note*: There does not need to exist a `System.ValueTuple<...>` declaration that directly matches the arity of any tuple type with a corresponding number of type parameters. Instead, larger tuples can be represented with a special overload `System.ValueTuple<..., TRest>` that in addition to tuple elements has a `Rest` field containing a nested tuple value of the remaining elements. Such nesting ...
```
This will be interesting to add to the relevant annex...

MadsTorgersen replied:
I added the `ValueTuple<...>` overloads that we actually use in .NET to the Annex. I hope that's acceptable.

jskeet replied:
Works for me :) (I wasn't sure whether we were going to try to allow any number of overloads to be available in an implementation.)

jskeet commented on standard/expressions.md:
```diff
@@ -1512,7 +1512,14 @@ tuple_element
-A tuple expression is classified as a tuple. It has a type if all the element expressions have a type. In that case, the type of a tuple expression is the tuple type of the same arity, where each type element has the type - and name, where given - of the corresponding element expression.
+A tuple expression is classified as a tuple. A tuple expression has a type if and only if each of its element expressions `Ei` has a type `Ti`. The type shall be a tuple type of the same arity as the tuple expression, where each element is given by the following:
+
+- If the tuple element in the corresponding position has a name `Ii`, then the tuple type element shall be `Ti Ii`.
```
Is there some other letter we could pick instead of `I` here? Maybe `Ni` instead of `Ii` (captain)?

MadsTorgersen replied:
Fair point. Changed to `Ni` in upcoming commit.

jskeet commented on standard/expressions.md:
```diff
@@ -1512,7 +1512,14 @@ tuple_element
+
+- If the tuple element in the corresponding position has a name `Ii`, then the tuple type element shall be `Ti Ii`.
+- Otherwise, if `Ei` is of the form `Ii` or `E.Ii` or `E?.Ii` then the tuple type element shall be `Ti Ii`, *unless* any of the following holds:
+  - Another element of the tuple expression has the name `Ii`,
```
"or" at the end of this bullet to match the one below?

MadsTorgersen replied:
Yup, fixed in upcoming commit

jskeet commented on standard/types.md:
```diff
@@ -389,7 +389,7 @@ An enumeration type is a distinct type with named constants. Every enumeration t
 ### §tuple-types-new-clause Tuple types
 
-A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...T ...
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two tuple type elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the ty ...
```
Nit: space after `***arity***.`

MadsTorgersen replied:
Fixed in upcoming commit.

jskeet commented on standard/expressions.md:
```diff
@@ -979,14 +979,14 @@ A function member is said to be an ***applicable function member*** with respect
 - Otherwise, the expanded form is applicable if for each argument in `A` the parameter-passing mode of the argument is identical to the parameter-passing mode of the corresponding parameter, and
   - for a fixed value parameter or a value parameter created by the expansion, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) exists from the argument expression to the type of the corresponding parameter, or
-  - for a `ref` or `out` parameter, the type of the argument expression is identical to the type of the corresponding parameter.
+  - for a `ref` or `out` parameter, the type of the argument expression is either implicit or identical to the type of the corresponding parameter.
```
By "identical" do we include identity conversions? (I'm sure we've had this conversation before...)

MadsTorgersen replied:
Yes! It should be the same as line 982 above. Fixed in upcoming commit.

jskeet reviewed (commented):
New commits look good. Just as with the patterns PR, I'm not a good judge of whether or not it's complete...

MadsTorgersen committed:
Fix declaration expressions and overload resolution

I've further unified `_` discards and declaration expressions.

Furthermore I've changed the approach to overload resolution. The change is small but radical, leaving out all `out` and `ref` parameters entirely from parameter-wise betterness!

I believe this is sound, since out and ref arguments are always required to match precisely, so one is never better than the other. However, the desired effect was just to ensure that declaration expressions do not influence betterness. If we don't like this fix we can reduce its scope to a bigger change with smaller consequences. ;-)

MadsTorgersen committed:
Root declaration_expression

Root declaration_expression as a non_assignment_expression and adjust verbiage accordingly in various places.

MadsTorgersen committed:
Add tuple equality

MadsTorgersen committed:
Fixing linter errors

MadsTorgersen committed:
Add deconstruction expressions

MadsTorgersen committed:
Fix lint errors

MadsTorgersen committed:
Fix wording

MadsTorgersen committed:
Root tuple_expression

MadsTorgersen committed:
Add definite assignment rules for new expression forms

Nigel-Ecma commented on standard/types.md:
```diff
@@ -389,36 +389,42 @@ An enumeration type is a distinct type with named constants. Every enumeration t
 ### §tuple-types-new-clause Tuple types
 
-A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***.A tuple type is written `(T1 I1, ..., Tn In)` with at least two elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...T ...
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***. A tuple type is written `(T1 I1, ..., Tn In)` with at least two tuple type elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the t ...
```
“at least two tuple type elements” could be read as two of the `Tn` must be tuple types?

```suggestion
A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***. A tuple type is written `(T1 I1, ..., Tn In)` with n ≥ 2, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the types `T1...Tn` from `System.ValueTuple<...>`, which shall be a set of generic struct types capable of expressing tuple types of any arity greater than one.
```

Nigel-Ecma commented on standard/types.md:
```diff
@@ -389,36 +389,42 @@ An enumeration type is a distinct type with named constants. Every enumeration t
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***. A tuple type is written `(T1 I1, ..., Tn In)` with at least two tuple type elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the t ...
 
-Element names within a tuple type shall be distinct. It is an error for an explicit tuple element name to be of the form `ItemX` where `X` is any sequence of non-`0`-initiated decimal digits that could represent the position of a tuple element, except as the name for that actual element.
+> *Note*: There does not need to exist a `System.ValueTuple<...>` declaration that directly matches the arity of any tuple type with a corresponding number of type parameters. Instead, larger tuples can be represented with a special overload `System.ValueTuple<..., TRest>` that in addition to tuple elements has a `Rest` field containing a nested tuple value of the remaining elements. Such nesting ...
```
This says an *n*-tuple in different implementations could be represented by a chain of *q* *m*-tuples and a final *p*-tuple where *n* = *q* × (*m*-1) + *p* and *m* is implementation dependent number. Do we need to say anything about how one implementation handles tuple values created by another with a different *m*?

MadsTorgersen replied:
Captured in #768 to be addressed post-merge.

Nigel-Ecma commented on standard/types.md:
```diff
@@ -389,36 +389,42 @@ An enumeration type is a distinct type with named constants. Every enumeration t
+A tuple type represents an ordered, fixed-length sequence of values with optional names and individual types. The number of elements in a tuple type is referred to as its ***arity***. A tuple type is written `(T1 I1, ..., Tn In)` with at least two tuple type elements, where the identifiers `I1...In` are optional ***tuple element names***. This syntax is shorthand for a type constructed with the t ...
 
-Element names within a tuple type shall be distinct. It is an error for an explicit tuple element name to be of the form `ItemX` where `X` is any sequence of non-`0`-initiated decimal digits that could represent the position of a tuple element, except as the name for that actual element.
+> *Note*: There does not need to exist a `System.ValueTuple<...>` declaration that directly matches the arity of any tuple type with a corresponding number of type parameters. Instead, larger tuples can be represented with a special overload `System.ValueTuple<..., TRest>` that in addition to tuple elements has a `Rest` field containing a nested tuple value of the remaining elements. Such nesting ...
```
The compiler usually constructs the underlying `ValueTuple<>` chain. If the program constructs an n-tuple directly using `ValueTuple<>` then is the code potentially non-portable between implementations which declare a different number of `ValueTuple<>` sizes? Or does the compiler accept only flat `new ValueTuple…`?

MadsTorgersen replied:
Captured in #768 to be addressed post-merge.

Nigel-Ecma commented on standard/variables.md:
```diff
@@ -152,20 +152,26 @@ A local variable introduced by a *local_variable_declaration* or *declaration_ex
-*Example:*
+> *Note*: `_` is a valid identifier in many forms of declarations. *end note*
+  
+Because a discard has no name, the only reference to the variable it represents is the expression that introduces it. A discard is not initially assigned, so it is always an error to access its value.
```
As the example below shows a discard can be passed as an `out` parameter, so within the called method the variable *must* be referenced as the parameter name is an alias for it. I know what this is trying to say but wordsmithing a suggestion escapes me at the moment.

MadsTorgersen replied:
I added a note; hopefully that's enough wordsmithing to get the point across.

MadsTorgersen committed:
Fix lint errors

MadsTorgersen commented:
> This is missing several special behaviors for tuples, e.g. the behavior of `==`.

The `==` and `!=` behavior is now specified.

MadsTorgersen committed:
wordsmith

Co-authored-by: Nigel-Ecma <perryresearch@zoot.net.nz>

MadsTorgersen committed:
Unify deconstruction expressions with tuple expresseions

MadsTorgersen committed:
Merge branch 'Mads-v7-local-declarations-in-expressions' of https://github.com/dotnet/csharpstandard into Mads-v7-local-declarations-in-expressions

MadsTorgersen commented on standard/expressions.md:
```diff
@@ -1480,6 +1499,70 @@ parenthesized_expression
+A *tuple_expression* is classified as a tuple.
+
+A *deconstruction_expression* `var (e1, ..., en)` is shorthand for the *tuple_expression* `(var e1, ..., var en)` and follows the same behavior. This applies recursively to any nested *deconstruction_tuple*s in the *deconstruction_expression*. Each identifier nested within a *deconstruction_expression* thus introduces a declaration expression (§declaration-expressions-new-clause). As a result, a  ...
+
```
An example would be really helpful

MadsTorgersen replied:
Captured in #768 to be addressed post-merge.

jskeet commented:
Agreed we're good to merge this - there are still tiny bits to do, but getting this in and then raising more bits will be better than trying to get it all the way.

MadsTorgersen committed:
Add note about discards as out arguments

MadsTorgersen committed:
Remove space
